proposals/readonly-methods.md

```diff
@@ -0,0 +1,65 @@
+# Readonly Methods
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Provide a way to specify individual methods on a struct do not modify state, in the same way that `readonly struct` specifies no instance method modifies state.
+
+It is worth noting that `readonly method` != `pure method`. A `pure` method guarantees no state will be modified. A `readonly` method only guarantees that instance state will not be modified.
+
+All instance methods on a `readonly struct` could be considered implicitly `readonly methods`. Explicit `readonly methods` declared on non-readonly structs would behave in the same manner. For example, they would still create hidden copies if you called an instance method (on the current instance or on a field of the instance) which was itself not-readonly.
+
+## Motivation
+[motivation]: #motivation
+
+Today, users have the ability to create `readonly struct` types which the compiler enforces that all fields are readonly (and by extension, that no methods modify the state). However, there are some scenarios where you have an existing API that exposes public fields, or needs to modify some small amount of internal state in select scenarios. Under these circumstances, you cannot mark the type as `readonly` (it would be a breaking change).
+
+This normally doesn't have much impact, except in the case of `in` parameters. With `in` parameters for non-readonly structs, the compiler will make a copy of the parameter for each instance method call, since it cannot guarantee that the call does not modify internal state. This can lead to a multitude of copies and worse overall performance than if you had just passed the struct directly by value. For an example, see this code on [sharplab](https://sharplab.io/#v2:CYLg1APgAgDABFAjAbgLACgNQMxwM4AuATgK4DGBcAagKYUD2RATBgN4ZycK4BmANvQCGlAB5p0XbnH5DKAT3GSOXHNIHC4AGRoA7AOYEAFgGUAjiUFEawZZ3YTJXPTQK3H9x54QB2OAAoROAAqOBEASjgwNy8YvzlguDkwxS8AXzd09EysXCgmOABhOA8VXnVKAFk/AEsdajoCRnyAN0E+EhoIks8oX1b2mgA6bX0jMwsrYEi4fo7h3QMTc0trFM5M1KA==)
+
+## Design
+[design]: #design
+
+I propose allowing a user to specify that a method itself is `readonly` and does not modify the state of the struct (with all the appropriate verification done by the compiler, of course). For example:
+
+```C#
+public struct Vector2
+{
+    public float x;
+    public float y;
+
+    public float GetLength() readonly
+    {
+        return MathF.Sqrt(LengthSquared);
+    }
+
+    public float LengthSquared
+    {
+        get readonly
+        {
+            return (x * x) +
+                   (y * y);
+        }
+    }
+}
+```
+
+The compiler would emit the method, as usual, and would additionally emit a compiler recognized attribute indicating that the method does not modify state.
+
+This would allow the user to safely call said instance method without the compiler needing to make a copy.
+
```

Comment:
Restrictions:

- The `readonly` modifier cannot be applied to `static` methods, constructors or destructors.
- The `readonly` modifier cannot be applied to `delegates`.
- The `readonly` modifier cannot be applied to members of `class` or `interface`.

Reply:
@jaredpar Why couldn't it be applied to members of `class` or `interface`? Not modifying the state of `this` is a valuable property. Or is it just a temporary restriction to limit the scope of the feature?

Reply:
@orthoxerox, we don't have `readonly class` today, so it isn't relevant.

If/when we get `readonly class`, then (assuming it was after) `readonly methods` would naturally be considered to extend there as well.

Reply:
@orthoxerox 

The use of `readonly` today provides a demonstrable benefit to developers: it avoids unnecessary `struct` copies. This can be a real benefit for high performance code. There is no equivalent motivating scenario for `class`, and `interface` to warrant adding this now.

Additionally I'm not sure it's the solution that we want for `class` and `interface`. There you likely want something closer to `const` in C++ (without all of the downsides like double method definitions). I have ideas on how to add notions of readonly behavior to reference types but it requires a bit more than just `readonly`.